[HUDI-3729][SPARK] fixed the per regression by enable vectorizeReader for parquet file

xiarixiaoyao opened the pull request:
… parquet file

## *Tips*
- *Thank you very much for contributing to Apache Hudi.*
- *Please review https://hudi.apache.org/contribute/how-to-contribute before opening a pull request.*

## What is the purpose of the pull request

*(For example: This pull request adds quick-start document.)*

## Brief change log

*(for example:)*
  - *Modify AnnotationLocation checkstyle rule in checkstyle.xml*

## Verify this pull request

*(Please pick either of the following options)*

This pull request is a trivial rework / code cleanup without any test coverage.

*(or)*

This pull request is already covered by existing tests, such as *(please describe tests)*.

(or)

This change added tests and can be verified as follows:

*(example:)*

  - *Added integration tests for end-to-end.*
  - *Added HoodieClientWriteTest to verify the change.*
  - *Manually verified the change by running a job locally.*

## Committer checklist

 - [ ] Has a corresponding JIRA in PR title & commit
 
 - [ ] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

xiarixiaoyao commented:
@xushiyan @leesf  @alexeykudinkin   pls  review this change
enable  vectorize  for cow/mor snapshort read. 
and benchmark test to prevent perf regression.

```
before change:
perf cow snapshot read:                   Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
vectorized disable                                 2204           2208           5          4.5         220.4       1.0X
vectorized enable                                  2183           2218          49          4.6         218.3       1.0X


after change:
perf cow snapshot read:                   Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
vectorized disable                                 2178           2180           2          4.6         217.8       1.0X
vectorized enable                                   659            674          24         15.2          65.9       3.3X
```

xushiyan commented on hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/MergeOnReadIncrementalRelation.scala:
```diff
@@ -48,6 +48,13 @@ class MergeOnReadIncrementalRelation(sqlContext: SQLContext,
   override type FileSplit = HoodieMergeOnReadFileSplit
 
+  override def imbueConfigs(sqlContext: SQLContext): Unit = {
+    sqlContext.sparkSession.sessionState.conf.setConfString("spark.sql.parquet.filterPushdown", "true")
+    sqlContext.sparkSession.sessionState.conf.setConfString("spark.sql.parquet.recordLevelFilter.enabled", "true")
```
why do we need to enforce these two as well?  `enableVectorizedReader` is the only one matters here right?

xiarixiaoyao replied:
mor  incremental  query  need to filter data through file level filters
from spark side
```
  val PARQUET_RECORD_FILTER_ENABLED = buildConf("spark.sql.parquet.recordLevelFilter.enabled")
    .doc("If true, enables Parquet's native record-level filtering using the pushed down " +
      "filters. " +
      s"This configuration only has an effect when '${PARQUET_FILTER_PUSHDOWN_ENABLED.key}' " +
      "is enabled and the vectorized reader is not used. You can ensure the vectorized reader " +
      s"is not used by setting '${PARQUET_VECTORIZED_READER_ENABLED.key}' to false.")
    .version("2.3.0")
    .booleanConf
    .createWithDefault(false)
```
We must ensure that these two configuration items take effect, otherwise the read data will be repeated

i also post another pr https://github.com/apache/hudi/pull/5165 to enable vectorize read for all cow/mor read

xushiyan replied:
Ok got it. thanks for the clarification

xushiyan reviewed (commented):
thanks for the patch the benchmarking! have 1 question below

nsivabalan commented:
👍 on the perf nos.

xiarixiaoyao commented on hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieBaseRelation.scala:
```diff
@@ -290,11 +290,8 @@ abstract class HoodieBaseRelation(val sqlContext: SQLContext,
-    // TODO(HUDI-3639) vectorized reader has to be disabled to make sure MORIncrementalRelation is working properly
-    sqlContext.sparkSession.sessionState.conf.setConfString("spark.sql.parquet.enableVectorizedReader", "false")
+  protected def imbueConfigs(sqlContext: SQLContext): Unit = {
+    // Nothing to do
```
ok， let me fixed it

xiarixiaoyao commented:
@alexeykudinkin  addressed the comment
but i still have a I have a question， do we really need force  set spark.sql.parquet.recordLevelFilter.enabled=true to mor/cow snapshot query
test prepare
set spark.sql.parquet.enableVectorizedReader=false， since spark.sql.parquet.recordLevelFilter.enabled is conflict with it.
here is the benchmark result and bench mark code：
```
        prepareHoodieCowTable(tableName, new Path(f.getCanonicalPath, tableName).toUri.toString)
        val benchmark = new HoodieBenchmark("perf cow snapshot read", 1000000)
        benchmark.addCase("recordLevelFilter enable") { _ =>
          // spark.sessionState.conf.setConfString("spark.sql.parquet.enableVectorizedReader", "false")
          spark.sessionState.conf.setConfString("spark.sql.parquet.filterPushdown", "true")
          spark.sessionState.conf.setConfString("spark.sql.parquet.recordLevelFilter.enabled", "true")
          spark.sql(s"select c1, c3, c4, c5 from $tableName").where("c1 > 100000").count()
        }
        benchmark.addCase("recordLevelFilter disable") { _ =>
          spark.sessionState.conf.setConfString("spark.sql.parquet.filterPushdown", "true")
          spark.sessionState.conf.setConfString("spark.sql.parquet.recordLevelFilter.enabled", "false")
          spark.sql(s"select c1, c3, c4, c5 from $tableName").where("c1 > 100000").count()
        }

perf cow snapshot read:                   Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
recordLevelFilter enable                            693            751          69          1.4         693.1       1.0X
recordLevelFilter disable                           662            680          27          1.5         662.4       1.0X

```
I don't see any performance improvement by set  spark.sql.parquet.recordLevelFilter.enabled=true

xiarixiaoyao commented:
kindly ping @alexeykudinkin  could you pls help me review again，  thank you very much

alexeykudinkin commented:
@xiarixiaoyao it depends on the data distribution in your table: filter push down would have an impact for ex, when it could rule out whole file reading column stats from Parquet footer.

danny0405 commented on hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/spark/sql/execution/benchmark/CowTableReadBenchmark.scala:
```diff
@@ -0,0 +1,128 @@
+        benchmark.addCase("vectorized disable") { _ =>
+          spark.sessionState.conf.setConfString("spark.sql.parquet.enableVectorizedReader", "false")
+          spark.sql(s"select c1, c3, c4, c5 from $tableName").count()
+        }
```
Maybe we can past the current benchmark result into the code comments so that we can diff with it for following releases.

xiarixiaoyao replied:
yes, of cource, let me do it

danny0405 reviewed (approved):
+1, nice job !

xiarixiaoyao commented:
UT passed, merge to the master

cxzl25 commented on hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/MergeOnReadSnapshotRelation.scala:
```diff
@@ -56,6 +56,11 @@ class MergeOnReadSnapshotRelation(sqlContext: SQLContext,
 
+  override def imbueConfigs(sqlContext: SQLContext): Unit = {
+    super.imbueConfigs(sqlContext)
+    sqlContext.sparkSession.sessionState.conf.setConfString("spark.sql.parquet.enableVectorizedReader", "true")
```
If we read the incremental and snapshot tables at the same time, the current configuration may result in inaccurate incremental data.
Because after enabling Vectorized Reader, parquet's recordLevelFilter will not work, and incremental query depends on recordLevelFilter to work, which may cause incremental query to generate a lot of irrelevant data.

Can we provide a configuration to decide whether to enable Vectorized Reader?

danny0405 replied:
I agree, we should make this option configurable.

danny0405 replied:
Or better, just disable vec reader when query type is incremental.
